syntax = "proto3";

package suggestions;

service SuggestionsService {
  rpc CalculateSuggestions(CalculateSuggestionsRequest) returns (CalculateSuggestionsResponse);
}

message Book {
  string book_id = 1;
  string title = 2;
  string author = 3;
}

message CalculateSuggestionsResponse {
  repeated Book suggested_books = 1;
}

message CalculateSuggestionsRequest {
  User user = 1;
  CreditCard credit_card = 2;
  string user_comment = 3;
  repeated Item items = 4;
  string discount_code = 5;
  string shipping_method = 6;
  string gift_message = 7;
  Address billing_address = 8;
  bool gift_wrapping = 9;
  bool terms_and_conditions_accepted = 10;
  repeated string notification_preferences = 11;
  Device device = 12;
  Browser browser = 13;
  string app_version = 14;
  string screen_resolution = 15;
  string referrer = 16;
  string device_language = 17;
}

message User {
  string name = 1;
  string contact = 2;
}

message CreditCard {
  string number = 1;
  string expiration_date = 2;
  string cvv = 3;
}

message Item {
  string name = 1;
  int32 quantity = 2;
}

message Address {
  string street = 1;
  string city = 2;
  string state = 3;
  string zip = 4;
  string country = 5;
}

message Device {
  string type = 1;
  string model = 2;
  string os = 3;
}

message Browser {
  string name = 1;
  string version = 2;
}




